akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -201,37 +226,42 @@ private[remote] final class HeaderBuilderImpl(
   def outboundClassManifestCompression: CompressionTable[String] = _outboundClassManifestCompression
 
   override def setSenderActorRef(ref: ActorRef): Unit = {
-    _senderActorRefIdx = outboundActorRefCompression.compress(ref)
-    if (_senderActorRefIdx == -1) _senderActorRef = Serialization.serializedActorPath(ref) // includes local address from `currentTransportInformation`
+    // serializedActorPath includes local address from `currentTransportInformation`
+    if (_useOutboundCompression) {
+      _senderActorRefIdx = outboundActorRefCompression.compress(ref)
+      if (_senderActorRefIdx == -1) _senderActorRef = Serialization.serializedActorPath(ref)
+    } else
+      _senderActorRef = Serialization.serializedActorPath(ref)
   }
   override def setNoSender(): Unit = {
     _senderActorRef = null
-    _senderActorRefIdx = EnvelopeBuffer.DeadLettersCode
+    _senderActorRefIdx = DeadLettersCode
   }
   override def isNoSender: Boolean =
-    (_senderActorRef eq null) && _senderActorRefIdx == EnvelopeBuffer.DeadLettersCode
+    (_senderActorRef eq null) && _senderActorRefIdx == DeadLettersCode
   override def senderActorRef(originUid: Long): OptionVal[ActorRef] =
-    if (_senderActorRef eq null)
+    if ((_senderActorRef eq null) && !isNoSender)
```

Comment:
maybe comment that for deadletters it is faster to avoid the table at all and treat it as an "always present" entry which does not need an explicit table lookup.